### PR TITLE
Rust: More information about extractor errors and warnings

### PR DESCRIPTION
**Pull request by geoffw0**
Output more information about extractor errors and warnings:
- add `rust/diagnostics/extraction-warnings` listing extractor warnings.
- update `rust/diagnostics/extraction-errors` to output only errors, not errors and warnings.
  - note that at the moment the extractor doesn't currently have any path to output a `Diagnostic` that is not an warning.  However that may change, and doing all this right ensures that code and DCA output are not confusing.
- add extraction errors, warnings, and successfully / unsuccessfully extracted files to `rust/summary/summary-statistics`.  This is helpful for examining databases manually.
- define a class `SuccessfullyExtractedFile` to improve code re-use.

### CODE REVIEW

File: rust/ql/lib/codeql/rust/Diagnostics.qll [hunk 1]
```diff
@@ -48,7 +48,12 @@ class Diagnostic extends @diagnostic {
   string toString() { result = this.getMessage() }
 }
 
-/** A diagnostic relating to a particular error in extracting a file. */
+/** A diagnostic that is error severity. */
 class ExtractionError extends Diagnostic {
-  ExtractionError() { this.getTag() = "parse_error" }
+  ExtractionError() { this.getSeverity() = 40 }
```
Comment (aibaars) [inline rust/ql/lib/codeql/rust/Diagnostics.qll:53]: Don't use `30` and `40`, use `@diagnostic_warning` and `@diagnostic_error` instead.

File: rust/ql/test/query-tests/diagnostics/SummaryStats.expected [hunk 2]
```diff
@@ -1,5 +1,9 @@
 | Elements extracted | 210 |
 | Elements unextracted | 0 |
-| Files extracted | 6 |
+| Extraction errors | 6 |
+| Extraction warnings | 0 |
```
Comment (aibaars) [inline rust/ql/test/query-tests/diagnostics/SummaryStats.expected:4]: I would expect `6` warnings and `0` errors, why is it the other way around in the summary?

File: rust/ql/lib/codeql/rust/Diagnostics.qll (hunk 1 above)
Comment (geoffw0) [reply rust/ql/lib/codeql/rust/Diagnostics.qll]: Yep that's much nicer, thanks.

File: rust/ql/test/query-tests/diagnostics/SummaryStats.expected (hunk 2 above)
Comment (geoffw0) [reply rust/ql/test/query-tests/diagnostics/SummaryStats.expected]: Ah, it looks like I hadn't updated the `.expected` file.  Now fixed.

File: rust/ql/src/queries/diagnostics/ExtractionWarnings.ql [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+/** Gets the SARIF severity to associate with a warning. */
+int getSeverity() { result = 1 }
+
+from ExtractionWarning warning, File f
+where
+  f = warning.getLocation().getFile() and
+  exists(f.getRelativePath())
+select warning, "Extraction warning in " + f + " with message " + warning.getMessage(),
+  getSeverity()
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply rust/ql/src/queries/diagnostics/ExtractionWarnings.ql:19]: The CPP version claims the "Extraction failed" with a warning.  Warnings don't cause our extractions to fail, so it's not the message we want.

**Comment by aibaars** [comment]
> @aibaars sync-identical-files is failing, it wants me to update the Ruby version of `Diagnostics.qll`. I think if I do that I may need to change some uses of `ExtractionError` as well, perhaps even give Ruby separate extraction errors / warnings queries. Or I can just turn the sync check off. How far do you (assuming you can speak for the Ruby team) think I should take this?

The purpose of the sync check is to avoid having diverging implementations between languages. I think we should strive for consistency across languages. It's really annoying if all the diagnostics libraries have very similar but slightly differing behaviour. 

I think your changes to Diagnostics.qll are fine, and could be applied to Ruby as well. Just make sure that behaviour doesn't change and add a change note if needed.

**Comment by geoffw0** [comment]
I've done the same changes for Ruby.  Definitely deserves a review from the Ruby perspective (and I wouldn't be surprised if some consistency tests need updating when CI has run them all).

**Comment by geoffw0** [comment]
@github/codeql-ruby please take a look at the Ruby changes here (that mirror changes made in Rust).  Thanks.

File: ruby/ql/lib/codeql/files/FileSystem.qll [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 private import codeql.Locations
 private import codeql.util.FileSystem
+private import codeql.ruby.Diagnostics
```
Comment (aibaars) [inline ruby/ql/lib/codeql/files/FileSystem.qll:5]: This QL module is language agnostic, try not to import any Ruby stuff. Perhaps these predicates are better placed in [`class RubyFile`](https://github.com/github/codeql/blob/25b3d76f71ae84abe8287814d27d280145ab070c/ruby/ql/lib/codeql/ruby/AST.qll#L139)

Although perhaps we should consider `Diagnostics` to be language neutral. Reporting extraction warnings on non-Ruby files, such as xml, json, yaml would actually make sense.

**Review by aibaars — DISMISSED** [review]
Looks good to me.

File: ruby/ql/lib/codeql/files/FileSystem.qll (hunk 4 above)
Comment (geoffw0) [reply ruby/ql/lib/codeql/files/FileSystem.qll]: I've moved the class to the same file as `class RubyFile`.  I'm not sure if you were saying you'd like it to be converted to a predicate as well.

> Although perhaps we should consider `Diagnostics` to be language neutral.

I can see some benefits to making `Diagnostics` language neutral and shared.  Way beyond the scope of this PR though.

File: ruby/ql/lib/codeql/ruby/AST.qll [hunk 5]
```diff
@@ -166,3 +167,20 @@ class RubyFile extends File {
   /** Gets the number of lines of comments in this file. */
   int getNumberOfLinesOfComments() { result = count(int line | this.line(line, true)) }
 }
+
+/**
+ * A successfully extracted file, that is, a file that was extracted and
+ * contains no extraction errors or warnings.
+ */
+class SuccessfullyExtractedFile extends RubyFile {
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/AST.qll:175]: If you don't want to restrict to Ruby source files only, you can do:
```suggestion
class SuccessfullyExtractedFile extends File {
```

Comment (geoffw0) [reply ruby/ql/lib/codeql/ruby/AST.qll]: Thanks, done.

**Review by aibaars — APPROVED** [review]

**Comment by geoffw0** [comment]
Both Rust and Ruby changes have been approved (admittedly by the same person), CI is happy.  Merging.

**Pull request merged**